cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ReentrantDistributedLock.java

```diff
@@ -0,0 +1,395 @@
+          completion.setException(t);
+        }
+      }
+    });
+
+    // Gets the result from the completion
+    try {
+      if (interruptible) {
+        localLockNode.set(completion.get(timeout, unit));
+      } else {
+        localLockNode.set(Uninterruptibles.getUninterruptibly(completion, timeout, unit));
+      }
+      return true;
+    } catch (InterruptedException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (TimeoutException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (CancellationException e) {
+      // If the completion get cancelled, meaning the lock acquisition is aborted.
+      return false;
+    }
+  }
+
+  /**
+   * Executes the lock acquisition process. This corresponds to step 2-5 of the distributed lock receipt.
+   */
+  private void doAcquire(final SettableFuture<String> completion, final boolean waitForLock,
+                         final String guid, @Nullable final String lockPath) {
+    // Step 2. Get all children under the lock parent.
+    Futures.addCallback(zkClient.getChildren(path), new FutureCallback<NodeChildren>() {
+
+      @Override
+      public void onSuccess(NodeChildren children) {
+        // Find the lock node in case the creation step failed by matching the guid
+        // See "Recoverable Errors and the GUID" in the ZooKeeper guide
+        final String lockNode = lockPath == null ? findLockNode(children.getChildren(), guid) : lockPath;
+        if (lockNode == null) {
+          // If not able to find the lock node, fail the locking procedure.
+          completion.setException(new IllegalStateException("Failed to acquire lock").fillInStackTrace());
+          return;
+        }
+
+        if (lockPath == null) {
+          // If lock node was not determined in step 1 due to connection loss exception, need to add the
+          // node deletion handler in here after the actual lockNode is determined.
+          deleteNodeOnFailure(completion, lockNode);
+        }
+
+        // Find the node to watch, which is the one with the largest id that is smaller than currentId
+        // If the current id is the smallest one, nodeToWatch will be null
+        String nodeToWatch = findNodeToWatch(children, lockNode, guid);
+
+        // Step 3a. lockNode is the lowest, hence this become lock owner.
+        if (nodeToWatch == null) {
+          // Acquired lock
+          completion.set(lockNode);
+        } else if (!waitForLock) {
+          // This is for the case of tryLock() without timeout.
+          completion.cancel(true);
+        }
+        // If the lock acquisition is completed, due to whatever reason, we need to watch for any other nodes
+        if (completion.isDone()) {
+          return;
+        }
+
+        // Step 3b and 4. Call exists on the next lowest sequence ID
+        OperationFuture<Stat> existsFuture = zkClient.exists(nodeToWatch, new Watcher() {
+          @Override
+          public void process(WatchedEvent event) {
```

Comment:
The watch is needed because when the "exists" call finds that the `nodeToWatch` still exists (meaning someone else is still holding the lock), this process needs to wait for the removal of the `nodeToWatch` so that this watcher will get triggered (and supposing this will become the owner of the lock, since it has the lowest seq now).

I think the logic can be simplified a bit by saying if the `nodeToWatch` is removed (as being told by the `WatchedEvent`), this lock becomes the owner (meaning no need to call `doAcquire` again).